HULKs/hulk
# Automatic calibration controller (Der dessen name nicht genannt werden darf II)

## Introduced Changes

The current work in #695 mainly provides the calibration line detection and extracting measurements (although the PR originally meant to implement the integration too).

- This PR will provide the behaviour and coordination components
    - Extracting measurements or running actual calibration is out of the scope
- ~#695 can be cleaned for calibration line detection and measurement extraction.~ -> Replaced by #1057

Whichever PR gets merged last, will connect the final integration (writing calibration results, validation, etc) or as a 3rd PR.

Depends on #1076 (for datatypes & workflow testing) - It can be decoupled if nessesary.
Fixes #131

State changes:
- Controller's commands (blue thin stripe is "capture"
- other two are vision cyclers, Yellow is "captured value" -> fake measurement generator randomly deciding to send or not.
![image](https://github.com/HULKs/hulk/assets/6742239/adc60f46-fec2-46be-8750-0633a0e9b3bc)

## Workflow

The controller node tracks the state of the primary state and issues the calibration commands when the primary state is set to Calibrate.

Commands & State

- *Inactive*
  - Do nothing when the primary state is anything other than `Calibrate`
  - Transition to Initialize when the primary state goes to `Calibrate` and also set the cycle time.
- *Initialize* `{ started time }`
  - Active during initialization, waits for `initial_stabilization_delay` until the robot finishes movement and stabilizes.
  - Once the timeout is reached, transition to LookAt with the stored LookAt command list. If nothing is found, jump to _Processing_ phase.
- *LookAt* `{ target, camera, dispatch_time }`
  - Move the head to the given target.
  - Waits `look_at_dispatch_waiting` from `dispatch_time` to ensure the head movement is complete.
  - Once the timeout is complete, switch to 'Capture'
- *Capture* `{ dispatch_time }`
  - The vision node extracts measurements using the camera images. Shall do necessary filtering in this phase as well.
  - Waits until a measurement arrives with this information & stores it.
      - This may need improvements - See TODOs
  - Transition to the next *LookAt* once capture information arrives or jump to Processing if all look-at commands are executed.
- *Processing*
  - Runs the solver, gets the results and persists to storage if satisfactory.
  - Transition to *Finish* once above is complete.
- *Finish*
  - End state. Does nothing until exiting calibration primary phase.
  - Auto-change primary state??

## ToDo / Known Issues

- [x] Verify that the use of `PerceptionInput` is correct
- [x] Handle when a measurement cannot be obtained from a given position (retry?, move to next look-at?
    - vision retries and says `RetriesExceeded` if it can't find any measurement.
    - If both cameras fail to find a measurement, it'll goto the next one.
- [x] Consider where to run the actual calibration processing -> on the node, on a future or a separate cycler
    - Done on the controller as of now, seems to be fast enough*
- [x] CalibrationCommand (also tracks state) could be improved.
    - Further actions added for Next Iterations
- [x] LookAt commands use a timeout to decide the end of the movement. Consider alternatives
   - Won't consider anything else for this PR
- [x] The initial phase (after activating calibration) uses a timeout (instead of checking if the pose is reached and stable). Consider alternatives
   - Won't consider anything else for this PR
- [x] Remove logging and replace with `AdditionalOutputs`
- [x] Nested `match` and `if` blocks could be made more elegant/ flatter
    - Done as much as I can. Reviewer could help!

## Ideas for Next Iterations (Not This PR)

- Make look-at commands configurable (got a fixed list for now)
- Finalize the integration and include tests for the entire workflow.
- ~Should consider adding camera to the `capture` command.~ (already done)
- Decide on keeping the timeout for look-at or check for reaching the look-at status (joint value?)
- Decide on keeping the timeout for initial to calibration start - might want to look at IMU, etc values for stability?
- Should we automatically switch to `Initial` state after reaching `Finish` state in calibration?
    - https://github.com/HULKs/hulk/issues/1092

## How to Test

Webots or a NAO:
- On Twix, subscribe to:
    - `Control.main_outputs.calibration_command` to look at the calibration phase
    - `Control.main_outputs.primary_state` to verify the transition from `initial` to `calibration` (useful for Webots)
- Enter the calibration mode from Initial, (press `ctrl + shift + x` on webots for 1 second of "simulated time")
- If on webots, `pepsi run` will print the state changes and measurement collection on the terminal
